### PR TITLE
maryam - adagrams

### PR DESCRIPTION
**Pull request by marshi23**
# JS Adagrams
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
--- | ---
What patterns were you able to use from your Ruby knowledge to apply to JavaScript? | The higher picture of how coding logic works like creating methods and variables to solve problems.  
Did you need to use different strategies to find information online about JavaScript than you would for Ruby? | For the most part,  I used the same strategies for my searches but used different sources. 
What was a challenge you were able to overcome on this assignment? |  Overcoming the limitations of writing code as a beginner, I am still learning how to work with nesting functions and using that to make my code more efficient.
What is your favorite thing about learning a new programming language? | Learning the strengths and capabilities of a new language and discovering what  I can do with it that I could not do with the other languages I know.
What is your least favorite thing about learning a new programming language? | Not knowing the syntax and having to google everything. 
Do you have any recommendations on how we could improve this project for the next cohort? | I think because we had a similar project  in the past, it was great project to launch JavaScript week with.

### CODE REVIEW

File: src/adagrams.js [hunk 1]
```diff
@@ -1,6 +1,143 @@
...
+    ];
+
+    let random = function(items) {
+      return items[Math.floor(Math.random()*items.length)];
+    }
+
+    let drawn = [];
+    for (let i = 0; i < 10; i++) {
+      let letter = random(letterDistribution);
+      drawn.push(letter)
```
Comment (Hamled) [inline src/adagrams.js:40]: Technically because this method is maintaining the letter distribution _per letter-draw_. As a result, each time this loop iterates we have the same chance to draw any given letter -- so it's possible we could end up drawing Z ten times, even though there's only one Z in the `letterDistribution` array.

To complete this method's functionality you should also _remove_ each letter from the `letterDistribution` array after it has been drawn.

File: src/adagrams.js [hunk 2]
```diff
@@ -1,6 +1,143 @@
...
+    let letters_array = input.toUpperCase().split('')
+
+    letters_array.forEach(function(letter) {
+
+      if (!lettersInHand.includes(letter)) {
+        result = false;
+      } else {
+
+        let index_2 = lettersInHand.indexOf(letter)
+        lettersInHand.splice(index_2, 1);
```
Comment (Hamled) [inline src/adagrams.js:57]: I like this approach to tracking which letters in `lettersInHand` we've already "counted" when checking this word.

However, it has the side effect that we're removing things from `lettersInHand` as we loop through it, which might interfere with whatever code called this method (it might be assuming the array it passes for `lettersInHand` will be unmodified.

As such, we should be more cautious with how we use that variable, and instead create a copy of it that we can delete from using `splice`. The code to make that copy can be done at the top of the function:

```js
usesAvailableLetters(input, lettersInHand) {
  const handCopy = [...lettersInHand];
  // ... rest of the code
},
```

In the above code we make a copy of the `lettersInHand` variable and then assign it to a new `const`, and the rest of the code should be updated to use `handCopy` instead of `lettersInHand`.

**Comment by Hamled** [comment]
# JS Adagrams
## What We're Looking For

Feature | Feedback
--- | ---
**General** |
Answered comprehension questions | :white_check_mark:
Small commits with meaningful commit messages | :white_check_mark:
**Code Requirements** |
`drawLetters` method | :white_check_mark:
Uses appropriate data structure to store the letter distribution | :white_check_mark:
All tests for `drawLetters` pass | :white_check_mark:
`usesAvailableLetters` method | :white_check_mark:
All tests for `usesAvailableLetters` pass | :white_check_mark:
`scoreWord` method | :white_check_mark:
Uses appropriate data structure to store the letter scores | :white_check_mark:
All tests for `scoreWord` pass | :white_check_mark:
`highestScoreFrom` method | :white_check_mark:
Appropriately handles edge cases for tie-breaking logic | :white_check_mark:
All tests for `highestScoreFrom` pass | :white_check_mark:
**Overall** | Looks good!<br>I have a few minor comments on some subtle behaviors (things that instructors should have written tests for, actually!) but otherwise everything is solid.
